File: composite-exo-glove/cpp/src/imu.cpp
#include "composite_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>

namespace {

void print_exo_imu(const qnbot::ExoTelemetry& telemetry) {
    if (telemetry.payload.torso_imu) {
        std::cout << "exo torso acceleration=";
        example::print_values(telemetry.payload.torso_imu->acceleration_mps2);
        std::cout << '\n';
    }
    if (telemetry.payload.extra_imu) {
        std::cout << "exo extra acceleration=";
        example::print_values(telemetry.payload.extra_imu->acceleration_mps2);
        std::cout << '\n';
    }
}

} // namespace

int main(int argc, char** argv) {
    try {
        qnbot::Sdk sdk(
            example::composite_config(example::parse_port(argc, argv)));
        auto glove_domain = sdk.glove();
        auto exo_domain = sdk.exo();
        auto glove = glove_domain.device("glove");
        auto exo = exo_domain.device("exo");
        const auto glove_subscription = glove.imu().subscribe(
            [](const qnbot::Sample<qnbot::GloveImu>& sample) {
                const auto& value = sample.value.payload;
                std::cout << "glove imu valid=" << value.valid << " gyro=["
                          << value.gyroscope_raw[0] << ", "
                          << value.gyroscope_raw[1] << ", "
                          << value.gyroscope_raw[2] << "]\n";
            });
        const auto exo_subscription = exo.telemetry().subscribe(
            [](const qnbot::Sample<qnbot::ExoTelemetry>& sample) {
                print_exo_imu(sample.value);
            });

        glove_domain.start();
        exo_domain.start();
        std::cout << "reading Glove and Exo IMU data; press Ctrl+C to stop\n";
        // Drive both member domains through the aggregate runner.
        sdk.run_forever();
        static_cast<void>(glove_subscription);
        static_cast<void>(exo_subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: composite-exo-glove/python/src/imu.py
from __future__ import annotations

import argparse

from qnbot_sdk import CompositeExoGloveConfig, Sample, Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig, ExoTelemetry
from qnbot_sdk.glove import GloveConfig, GloveImu

EXO_NAME = "exo"
GLOVE_NAME = "glove"


def create_sdk(port: str) -> Sdk:
    return Sdk(
        devices=(
            CompositeExoGloveConfig(
                connection=SerialConnection(port=port),
                devices=(GloveConfig(name=GLOVE_NAME), ExoConfig(name=EXO_NAME)),
            ),
        )
    )


def print_glove_imu(sample: Sample[GloveImu]) -> None:
    imu = sample.value.payload
    print(
        f"glove imu sequence={sample.sequence} valid={imu.valid} "
        f"gyro={imu.gyroscope_raw} accel={imu.accelerometer_raw}"
    )


def print_exo_imu(sample: Sample[ExoTelemetry]) -> None:
    payload = sample.value.payload
    for label, imu in (("torso", payload.torso_imu), ("extra", payload.extra_imu)):
        if imu is not None:
            print(
                f"exo {label} imu sequence={sample.sequence} "
                f"acceleration={imu.acceleration_mps2} "
                f"angular_velocity={imu.angular_velocity_rad_s}"
            )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Read Glove and Exo IMU data")
    arguments.add_argument("--port", required=True, help="Shared serial port")
    options = arguments.parse_args()

    sdk = create_sdk(port=options.port)
    try:
        glove_domain = sdk.glove()
        exo_domain = sdk.exo()
        glove = glove_domain.device(GLOVE_NAME)
        exo = exo_domain.device(EXO_NAME)
        glove.imu().subscribe(print_glove_imu)
        exo.telemetry().subscribe(print_exo_imu)
        glove_domain.start()
        exo_domain.start()
        print("running; press Ctrl+C to stop", flush=True)
        # The aggregate runner is required to drive both member domains.
        sdk.run_forever()
    except KeyboardInterrupt:
        glove_domain.stop()
        exo_domain.stop()
    finally:
        glove_domain.close()
        exo_domain.close()


if __name__ == "__main__":
    main()
